aws-s3-transfer-manager/tests/e2e_transfer_test.rs

```diff
@@ -0,0 +1,298 @@
+            &composite_key,
+            ChecksumStrategy::with_calculated_crc32_composite_if_multipart(),
+            file_size,
+        )
+        .await;
+
+        if is_multi_part {
+            // A multipart upload, so that the composite checksum should ends with `-<number of parts>`
+            let num_parts = (size_mb as f64 / 8f64).ceil() as u32;
+            assert!(composite_checksum.ends_with(&format!("-{}", num_parts)));
+        } else {
+            // A single object upload, so that the composite checksum should not contain `-` and matches the checksum before
+            assert!(!composite_checksum.contains('-'));
+            assert_eq!(composite_checksum, checksum);
+        }
+    }
+}
+
+#[tokio::test]
+async fn test_multi_part_file_checksum_upload() {
+    run_checksum_test(20, "20MB-crc32", true).await;
+}
+
+#[tokio::test]
+async fn test_single_part_file_checksum_upload() {
+    run_checksum_test(1, "1MB-crc32", false).await;
+}
+
+// TODO: add checksum validation tests for get object
+
+#[derive(Debug)]
+struct TestStream {
+    idx: usize,
+    remaining: usize,
+    delay: Option<Pin<Box<Sleep>>>,
+}
+
+impl TestStream {
+    fn new_delay(total_size: usize) -> Self {
+        let delay = Box::pin(tokio::time::sleep(Duration::from_secs(5)));
+        Self {
+            idx: 0,
+            remaining: total_size,
+            delay: Some(delay),
+        }
+    }
+}
+
+impl PartStream for TestStream {
+    fn poll_part(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        stream_cx: &StreamContext,
+    ) -> Poll<Option<std::io::Result<PartData>>> {
+        let part_size = stream_cx.part_size();
+        // Check if we need to wait for the delay
+        if let Some(delay) = &mut self.delay {
+            match delay.as_mut().poll(cx) {
+                Poll::Ready(_) => self.delay = None,
+                Poll::Pending => return Poll::Pending,
+            }
+        }
+
+        if self.remaining == 0 {
+            // End of stream
+            Poll::Ready(None)
+        } else {
+            let part_data_size = std::cmp::min(part_size, self.remaining);
+            let data = "This is a test"
+                .bytes()
+                .cycle()
+                .take(part_data_size)
+                .collect::<Vec<u8>>();
+            let part = PartData::new((self.idx + 1) as u64, data);
+
+            // Update state
+            self.idx += 1;
+            self.remaining -= part_data_size;
+            // Schedule delay for NEXT part
+            self.delay = Some(Box::pin(tokio::time::sleep(Duration::from_secs(3))));
+            Poll::Ready(Some(Ok(part)))
+        }
+    }
+
+    fn size_hint(&self) -> SizeHint {
+        SizeHint::exact(self.remaining as u64)
+    }
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 3)]
+async fn test_upload_with_long_running_stream() {
+    let (tm, _) = test_tm().await;
+    let file_size = 10 * 1024 * 1024; // 10MB
```

Comment:
We have helpers for this like https://github.com/awslabs/aws-s3-transfer-manager-rs/blob/main/aws-s3-transfer-manager/tests/download_test.rs#L463

Reply:
it's even more typing...